esiclient/tests/functional/base.py

```diff
@@ -0,0 +1,188 @@
+        admin_client = self.clients['admin']
+
+        # NOTE: names created using tempest's data_utils.rand_name() function
+        #       will be prefixed with 'tempest-' and suffixed with randomly
+        #       generated characters to prevent name collision.
+        project_name = data_utils.rand_name('esi-%s-project' % name)
+        flags = '--domain default --enable -f json'
+        if parent:
+            if parent not in self.projects.keys():
+                raise NameError('Invalid parent project: %s' % parent)
+            flags += ' --parent %s' % self.projects[parent]['name']
+        output = admin_client.openstack('project create %s' % project_name,
+                                        '', flags)
+        self._cls_cleanups.append([admin_client.openstack, 'project delete',
+                                   '', '%s' % project_name])
+        project_id = json.loads(output)['id']
+        self.projects[name] = {
+            'name': project_name,
+            'id': project_id,
+            'parent': parent,
+            'domain': 'default'
+        }
+
+        if roles is []:
+            raise ValueError('No roles specified when initializing dummy \
+                              project %s' % name)
+        elif type(roles) is str:
+            roles = [roles]
+
+        for role in roles:
+
+            username = data_utils.rand_name('esi-%s-%s' % (name, role))
+            password = data_utils.rand_password()
+            output = admin_client.openstack('user create %s' % username, '',
+                                            '--domain default --password %s \
+                                             --enable -f json' % password)
+            self._cls_cleanups.append([admin_client.openstack, 'user delete',
+                                       '', '%s' % username])
+            user_id = json.loads(output)['id']
+
+            client_name = '%s-%s' % (name, role)
+            self.users[client_name] = {
+                'name': username,
+                'id': user_id,
+                'password': password,
+                'domain': 'default'
+            }
+
+            admin_client.openstack('role add', '', '--user %s --project %s %s'
+                                   % (username, project_name, role))
+            self._cls_cleanups.append([admin_client.openstack, 'role remove',
+                                       '', '--user %s --project %s %s' %
+                                       (username, project_name, role)])
+
+            self._init_client(self, client_name, name)
+
+    def _verify_roles(self):
+        # Checks to see the is_node_owner and is_node_lessee roles exist
+        for role in 'owner', 'lessee':
+            self.clients['admin'].openstack('role show', '', '%s' % role)
+
+
+class CLIClient(base.CLIClient):
+
+    def ironic(self, action, flags='', params='', fail_ok=False,
+               merge_stderr=False):
+        return self.openstack('ironic %s' % action, '',
```

Comment:
Do CLI commands of the form `$ openstack ironic [...]` exist? I was under the impression they were all of the form `$ openstack baremetal [...]`. If not, this method should probably be changed to `CLIClient.baremetal(...)` and if i'm wrong about this, a `CLIClient.baremetal(...)` method should probably be added.

Reply:
This code is wrong, but I dont use it anywhere. But youre right. I ended up preferring to have a more explicit function for each call we make, as that was how Ironic did things in their code.